twine/auth.py

```diff
@@ -0,0 +1,80 @@
+import warnings
+import getpass
+import functools
+
+import keyring
+import munch
+
+from . import utils
+from . import exceptions
+
+
+class Resolver:
+    def __init__(self, settings, username=None, password=None):
+        self.settings = settings
+        self.input = munch.Munch(username=username, password=password)
```

Comment:
What about making this a [typed named tuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple)?

```python
class Credentialnput(NamedTuple):
    username: Optional[str]
    password: Optional[str]

self.input = CredentialInput(username, password)
```

Reply:
It's a little annoying to have to declare a type just to get a namespace. I chose `Munch` because it fits nicely on one line. I had considered `namedtuple`, but the syntax is distracting:

```
self.input = collections.namedtuple('CredentialInput', 'username password')(username, password)
```

`Munch` is more intuitive to read.

The typed named tuple is even more verbose, requiring 4 lines to express what Munch does in 1, but I guess if we're moving to the world of declaring types for everything, that's the best approach.

I definitely like the named tuples for their immutability and non-item access (`Munch` has a lot of dict-like behavior that I didn't want). I'll go with this suggestion.

Reply:
Implemented in 33d1365631d2573ca8835e865e4057131f184218.

I started out by simply implementing this type in the auth module, but then realized that the presence of this class makes for a good wrapper around the username/password pair, so extracted it into its own module (types), simplifying the signature of an auth.Resolver even more.